{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Curating Read Until input files for MinKNOW",
      "provenance": [],
      "authorship_tag": "ABX9TyM+MtgL1gNRfz4Hcn5/sIwN",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "6258c883fba94111aac11f787eb3804d": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DropdownModel",
          "state": {
            "_options_labels": [
              "--",
              "acanthochromis_polyacanthus",
              "accipiter_nisus",
              "ailuropoda_melanoleuca",
              "amazona_collaria",
              "amphilophus_citrinellus",
              "amphiprion_ocellaris",
              "amphiprion_percula",
              "anabas_testudineus",
              "anas_platyrhynchos",
              "anas_platyrhynchos_platyrhynchos",
              "anolis_carolinensis",
              "anser_brachyrhynchus",
              "anser_cygnoides",
              "aotus_nancymaae",
              "apteryx_haastii",
              "apteryx_owenii",
              "apteryx_rowi",
              "aquila_chrysaetos_chrysaetos",
              "astatotilapia_calliptera",
              "astyanax_mexicanus",
              "astyanax_mexicanus_pachon",
              "athene_cunicularia",
              "betta_splendens",
              "bison_bison_bison",
              "bos_grunniens",
              "bos_indicus_hybrid",
              "bos_mutus",
              "bos_taurus",
              "bos_taurus_hybrid",
              "caenorhabditis_elegans",
              "calidris_pugnax",
              "calidris_pygmaea",
              "callithrix_jacchus",
              "callorhinchus_milii",
              "camarhynchus_parvulus",
              "camelus_dromedarius",
              "canis_lupus_dingo",
              "canis_lupus_familiaris",
              "canis_lupus_familiarisbasenji",
              "canis_lupus_familiarisgreatdane",
              "capra_hircus",
              "carassius_auratus",
              "carlito_syrichta",
              "castor_canadensis",
              "catagonus_wagneri",
              "cavia_aperea",
              "cavia_porcellus",
              "cebus_capucinus",
              "cercocebus_atys",
              "chelonoidis_abingdonii",
              "chelydra_serpentina",
              "chinchilla_lanigera",
              "chlorocebus_sabaeus",
              "choloepus_hoffmanni",
              "chrysemys_picta_bellii",
              "chrysolophus_pictus",
              "ciona_intestinalis",
              "ciona_savignyi",
              "clupea_harengus",
              "colobus_angolensis_palliatus",
              "cottoperca_gobio",
              "coturnix_japonica",
              "cricetulus_griseus_chok1gshd",
              "cricetulus_griseus_crigri",
              "cricetulus_griseus_picr",
              "crocodylus_porosus",
              "cyanistes_caeruleus",
              "cynoglossus_semilaevis",
              "cyprinodon_variegatus",
              "cyprinus_carpio",
              "cyprinus_carpio_germanmirror",
              "cyprinus_carpio_hebaored",
              "cyprinus_carpio_huanghe",
              "danio_rerio",
              "dasypus_novemcinctus",
              "delphinapterus_leucas",
              "denticeps_clupeoides",
              "dicentrarchus_labrax",
              "dipodomys_ordii",
              "dromaius_novaehollandiae",
              "drosophila_melanogaster",
              "echeneis_naucrates",
              "echinops_telfairi",
              "electrophorus_electricus",
              "eptatretus_burgeri",
              "equus_asinus_asinus",
              "equus_caballus",
              "erinaceus_europaeus",
              "erpetoichthys_calabaricus",
              "erythrura_gouldiae",
              "esox_lucius",
              "felis_catus",
              "ficedula_albicollis",
              "fukomys_damarensis",
              "fundulus_heteroclitus",
              "gadus_morhua",
              "gallus_gallus",
              "gambusia_affinis",
              "gasterosteus_aculeatus",
              "geospiza_fortis",
              "gopherus_agassizii",
              "gopherus_evgoodei",
              "gorilla_gorilla",
              "gouania_willdenowi",
              "haplochromis_burtoni",
              "heterocephalus_glaber_female",
              "heterocephalus_glaber_male",
              "hippocampus_comes",
              "homo_sapiens",
              "hucho_hucho",
              "ictalurus_punctatus",
              "ictidomys_tridecemlineatus",
              "jaculus_jaculus",
              "junco_hyemalis",
              "kryptolebias_marmoratus",
              "labrus_bergylta",
              "larimichthys_crocea",
              "lates_calcarifer",
              "laticauda_laticaudata",
              "latimeria_chalumnae",
              "lepidothrix_coronata",
              "lepisosteus_oculatus",
              "lonchura_striata_domestica",
              "loxodonta_africana",
              "lynx_canadensis",
              "macaca_fascicularis",
              "macaca_mulatta",
              "macaca_nemestrina",
              "manacus_vitellinus",
              "mandrillus_leucophaeus",
              "marmota_marmota_marmota",
              "mastacembelus_armatus",
              "maylandia_zebra",
              "meleagris_gallopavo",
              "melopsittacus_undulatus",
              "meriones_unguiculatus",
              "mesocricetus_auratus",
              "microcebus_murinus",
              "microtus_ochrogaster",
              "mola_mola",
              "monodelphis_domestica",
              "monopterus_albus",
              "moschus_moschiferus",
              "mus_caroli",
              "mus_musculus",
              "mus_musculus_129s1svimj",
              "mus_musculus_aj",
              "mus_musculus_akrj",
              "mus_musculus_balbcj",
              "mus_musculus_c3hhej",
              "mus_musculus_c57bl6nj",
              "mus_musculus_casteij",
              "mus_musculus_cbaj",
              "mus_musculus_dba2j",
              "mus_musculus_fvbnj",
              "mus_musculus_lpj",
              "mus_musculus_nodshiltj",
              "mus_musculus_nzohlltj",
              "mus_musculus_pwkphj",
              "mus_musculus_wsbeij",
              "mus_pahari",
              "mus_spicilegus",
              "mus_spretus",
              "mustela_putorius_furo",
              "myotis_lucifugus",
              "myripristis_murdjan",
              "nannospalax_galili",
              "neogobius_melanostomus",
              "neolamprologus_brichardi",
              "neovison_vison",
              "nomascus_leucogenys",
              "notamacropus_eugenii",
              "notechis_scutatus",
              "nothoprocta_perdicaria",
              "numida_meleagris",
              "ochotona_princeps",
              "octodon_degus",
              "oncorhynchus_mykiss",
              "oncorhynchus_tshawytscha",
              "oreochromis_aureus",
              "oreochromis_niloticus",
              "ornithorhynchus_anatinus",
              "oryctolagus_cuniculus",
              "oryzias_javanicus",
              "oryzias_latipes",
              "oryzias_latipes_hni",
              "oryzias_latipes_hsok",
              "oryzias_melastigma",
              "oryzias_sinensis",
              "otolemur_garnettii",
              "ovis_aries",
              "pan_paniscus",
              "pan_troglodytes",
              "panthera_leo",
              "panthera_pardus",
              "panthera_tigris_altaica",
              "papio_anubis",
              "parambassis_ranga",
              "paramormyrops_kingsleyae",
              "parus_major",
              "pavo_cristatus",
              "pelodiscus_sinensis",
              "pelusios_castaneus",
              "periophthalmus_magnuspinnatus",
              "peromyscus_maniculatus_bairdii",
              "petromyzon_marinus",
              "phascolarctos_cinereus",
              "phasianus_colchicus",
              "physeter_catodon",
              "piliocolobus_tephrosceles",
              "podarcis_muralis",
              "poecilia_formosa",
              "poecilia_latipinna",
              "poecilia_mexicana",
              "poecilia_reticulata",
              "pogona_vitticeps",
              "pongo_abelii",
              "procavia_capensis",
              "prolemur_simus",
              "propithecus_coquereli",
              "pseudonaja_textilis",
              "pteropus_vampyrus",
              "pundamilia_nyererei",
              "pygocentrus_nattereri",
              "rattus_norvegicus",
              "rhinolophus_ferrumequinum",
              "rhinopithecus_bieti",
              "rhinopithecus_roxellana",
              "saccharomyces_cerevisiae",
              "saimiri_boliviensis_boliviensis",
              "salarias_fasciatus",
              "salmo_salar",
              "salmo_trutta",
              "salvator_merianae",
              "sarcophilus_harrisii",
              "scleropages_formosus",
              "scophthalmus_maximus",
              "serinus_canaria",
              "seriola_dumerili",
              "seriola_lalandi_dorsalis",
              "sinocyclocheilus_anshuiensis",
              "sinocyclocheilus_grahami",
              "sinocyclocheilus_rhinocerous",
              "sorex_araneus",
              "sparus_aurata",
              "spermophilus_dauricus",
              "sphaeramia_orbicularis",
              "sphenodon_punctatus",
              "stachyris_ruficeps",
              "stegastes_partitus",
              "strigops_habroptila",
              "struthio_camelus_australis",
              "suricata_suricatta",
              "sus_scrofa",
              "sus_scrofa_bamei",
              "sus_scrofa_berkshire",
              "sus_scrofa_hampshire",
              "sus_scrofa_jinhua",
              "sus_scrofa_landrace",
              "sus_scrofa_largewhite",
              "sus_scrofa_meishan",
              "sus_scrofa_pietrain",
              "sus_scrofa_rongchang",
              "sus_scrofa_tibetan",
              "sus_scrofa_usmarc",
              "sus_scrofa_wuzhishan",
              "taeniopygia_guttata",
              "takifugu_rubripes",
              "terrapene_carolina_triunguis",
              "tetraodon_nigroviridis",
              "theropithecus_gelada",
              "tupaia_belangeri",
              "tursiops_truncatus",
              "urocitellus_parryii",
              "ursus_americanus",
              "ursus_maritimus",
              "ursus_thibetanus_thibetanus",
              "varanus_komodoensis",
              "vicugna_pacos",
              "vombatus_ursinus",
              "vulpes_vulpes",
              "xenopus_tropicalis",
              "xiphophorus_couchianus",
              "xiphophorus_maculatus",
              "zonotrichia_albicollis",
              "zosterops_lateralis_melanops"
            ],
            "_view_name": "DropdownView",
            "style": "IPY_MODEL_1d00ce551a1749d3894348ff31fcd8bd",
            "_dom_classes": [],
            "description": "species:",
            "_model_name": "DropdownModel",
            "index": 1,
            "_view_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "disabled": false,
            "_view_module_version": "1.5.0",
            "description_tooltip": null,
            "_model_module": "@jupyter-widgets/controls",
            "layout": "IPY_MODEL_b3eb872b68cb4ed8b7657352f67a68e0"
          }
        },
        "1d00ce551a1749d3894348ff31fcd8bd": {
          "model_module": "@jupyter-widgets/controls",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_view_name": "StyleView",
            "_model_name": "DescriptionStyleModel",
            "description_width": "",
            "_view_module": "@jupyter-widgets/base",
            "_model_module_version": "1.5.0",
            "_view_count": null,
            "_view_module_version": "1.2.0",
            "_model_module": "@jupyter-widgets/controls"
          }
        },
        "b3eb872b68cb4ed8b7657352f67a68e0": {
          "model_module": "@jupyter-widgets/base",
          "model_name": "LayoutModel",
          "state": {
            "_view_name": "LayoutView",
            "grid_template_rows": null,
            "right": null,
            "justify_content": null,
            "_view_module": "@jupyter-widgets/base",
            "overflow": null,
            "_model_module_version": "1.2.0",
            "_view_count": null,
            "flex_flow": null,
            "width": null,
            "min_width": null,
            "border": null,
            "align_items": null,
            "bottom": null,
            "_model_module": "@jupyter-widgets/base",
            "top": null,
            "grid_column": null,
            "overflow_y": null,
            "overflow_x": null,
            "grid_auto_flow": null,
            "grid_area": null,
            "grid_template_columns": null,
            "flex": null,
            "_model_name": "LayoutModel",
            "justify_items": null,
            "grid_row": null,
            "max_height": null,
            "align_content": null,
            "visibility": null,
            "align_self": null,
            "height": null,
            "min_height": null,
            "padding": null,
            "grid_auto_rows": null,
            "grid_gap": null,
            "max_width": null,
            "order": null,
            "_view_module_version": "1.2.0",
            "grid_template_areas": null,
            "object_position": null,
            "object_fit": null,
            "grid_auto_columns": null,
            "margin": null,
            "display": null,
            "left": null
          }
        }
      }
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/epi2me-labs/tutorials/blob/master/Curating_Read_Until_input_files_for_MinKNOW.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0IsPU13jShte",
        "colab_type": "text"
      },
      "source": [
        "<h1>Curating Read Until input files for MinKNOW</h1>\n",
        "\n",
        "The following short workflow will prepare and download the necessary file to perform a Read Until sequence experiment selecting for reads that span genes, transcripts, exons, etc. stored within [ensembl](https://www.ensembl.org/).\n",
        "\n",
        "To prepare the files, simply execute teh code cells below in sequence by pressing the `Play` button to the left-hand side."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "rXy_YKVGP_OU",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 68
        },
        "cellView": "form",
        "outputId": "08f2b3da-c74d-40d9-8983-795159a8138c"
      },
      "source": [
        "#@markdown *Code installation*\n",
        "print(\" * Installing software\")\n",
        "!pip install pyranges > /dev/null\n",
        "from ftplib import FTP\n",
        "import os\n",
        "from google.colab import files\n",
        "import ipywidgets as widgets\n",
        "import pandas as pd\n",
        "import pyranges as pr\n",
        "import requests\n",
        "\n",
        "\n",
        "class EnsemblRestClient(object):\n",
        "    def __init__(self, server='http://rest.ensembl.org'):\n",
        "        self.server = 'http://rest.ensembl.org'\n",
        "        self.ftp = 'ftp.ensembl.org'\n",
        "        self.ftp_dna_path = '/pub/release-100/fasta/{}/dna/'\n",
        "        self.ftp_dna_suff = {\n",
        "            'primary':'dna.primary_assembly.fa.gz',\n",
        "            'toplevel':'dna.toplevel.fa.gz'}\n",
        "        self.ftp_gtf_path = '/pub/release-100/gtf/{}/'\n",
        "        self.ftp_gtf_suff = {\n",
        "            'gtf':\"100.gtf.gz\"}\n",
        "\n",
        "        self.dna_template = \\\n",
        "            \"ftp://\" + self.ftp + self.ftp_dna_path + \\\n",
        "            \"/{}.{}.\" + self.ftp_dna_suff['toplevel']\n",
        "        self.gtf_template = \\\n",
        "            \"ftp://\" + self.ftp + self.ftp_gtf_path + \\\n",
        "            \"/{}.{}.\" + self.ftp_gtf_suff['gtf']\n",
        "\n",
        "    def get(self, endpoint, params=dict(), **kwargs):\n",
        "        if 'json' not in kwargs:\n",
        "            kwargs['json'] = params\n",
        "        data = dict()\n",
        "        try:\n",
        "            response = requests.get(self.server + endpoint, **kwargs)\n",
        "            if response.status_code == 429:\n",
        "                if 'Retry-After' in response.headers:\n",
        "                    retry = e.headers['Retry-After']\n",
        "                    time.sleep(float(retry))\n",
        "                    response = requests.get(self.server + endpoint, **kwargs)\n",
        "        except:\n",
        "            print(' - Request failed for {0}'.format(endpoint))\n",
        "            print(response.status_code) \n",
        "        else:\n",
        "            data = response.json()\n",
        "            if \"error\" in data:\n",
        "                print(\" - ERROR:\\n   {}\".format(data[\"error\"]))\n",
        "        return data\n",
        "\n",
        "    def species_list(self):\n",
        "        return self.get(\"/info/species\")\n",
        "\n",
        "    def assembly_name(self, species):\n",
        "        #assembly = client.get('/info/assembly/{}'.format(species))\n",
        "        #if 'assembly_name' in assembly:\n",
        "        #    return assembly['assembly_name']\n",
        "        paths = self._ftp_list(\n",
        "            self.ftp_dna_path.format(species), self.ftp_dna_suff)\n",
        "        return paths['toplevel'].split('.')[1]\n",
        "\n",
        "    def _ftp_list(self, path, filt):\n",
        "        ftpdata = dict()\n",
        "        with FTP('ftp.ensembl.org') as ftp:\n",
        "            ftp.login()\n",
        "            def grab(x):\n",
        "                fname = x.split()[-1]\n",
        "                for key, value in filt.items():\n",
        "                    if fname.endswith(value):\n",
        "                        ftpdata[key] = fname\n",
        "            ftp.dir(path, grab)\n",
        "        return ftpdata\n",
        "\n",
        "    def dna_url(self, species, toplevel=True, assembly_name=None):\n",
        "        if assembly_name is None:\n",
        "            assembly_name = client.assembly_name(species)\n",
        "        return self.dna_template.format(\n",
        "            species, species.capitalize(), assembly_name)\n",
        "        \n",
        "    def gtf_url(self, species, assembly_name=None):\n",
        "        if assembly_name is None:\n",
        "            assembly_name = client.assembly_name(species)\n",
        "        return self.gtf_template.format(\n",
        "            species, species.capitalize(), assembly_name)\n",
        "\n",
        "\n",
        "print(\" * Querying ensembl species\")\n",
        "client = EnsemblRestClient()\n",
        "species_list = client.species_list()\n",
        "species_list = sorted(s['name'] for s in species_list['species'])\n",
        "print(\" - Found {} species\".format(len(species_list)))\n",
        "species_list.insert(0, \"--\")\n",
        "urls = (None, None)"
      ],
      "execution_count": 98,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            " * Installing software\n",
            " * Querying ensembl species\n",
            " - Found 286 species\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "QRg7J88GSi9W",
        "colab_type": "text"
      },
      "source": [
        "To efficiently produce reasonable target regions please provide an average read length. This should be an arithmetic mean not an N50 length. After pressing play here you will be given the opportunity to select your genome of interest from a drop-down box."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "giK73pqCEl9H",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 83,
          "referenced_widgets": [
            "6258c883fba94111aac11f787eb3804d",
            "1d00ce551a1749d3894348ff31fcd8bd",
            "b3eb872b68cb4ed8b7657352f67a68e0"
          ]
        },
        "cellView": "both",
        "outputId": "525cbd06-e291-4f69-9da2-2dd1b3e55ee5"
      },
      "source": [
        "read_length =  5000 #@param {type:\"integer\"}\n",
        "\n",
        "def species_change(change):\n",
        "    global urls\n",
        "    if change['type'] == 'change' and change['name'] == 'value':\n",
        "        print(\" * Finding files, please wait...\", end=\"\")\n",
        "        spec = change['new']\n",
        "        assm = client.assembly_name(spec)\n",
        "        dna_url = client.dna_url(spec, assembly_name=assm)\n",
        "        gtf_url = client.gtf_url(spec, assembly_name=assm)\n",
        "        urls = (dna_url, gtf_url)\n",
        "        print(\"done\")\n",
        "\n",
        "print(\"Select a species:\")\n",
        "species_dropdown = widgets.Dropdown(\n",
        "    options=species_list, value='--', description='species:')\n",
        "species_dropdown.observe(species_change)\n",
        "display(species_dropdown)"
      ],
      "execution_count": 99,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Select a species:\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "6258c883fba94111aac11f787eb3804d",
              "version_minor": 0,
              "version_major": 2
            },
            "text/plain": [
              "Dropdown(description='species:', options=('--', 'acanthochromis_polyacanthus', 'accipiter_nisus', 'ailuropoda_…"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            " * Finding files, please wait...done\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2gLaDwHBSkOb",
        "colab_type": "text"
      },
      "source": [
        "After the message `* Finding files, please wait...done` has been display above, press play on the next code cell to retrieve the required data and prepare the files required for MinKNOW.\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "R_YjqlKnPh6h",
        "colab_type": "code",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 374
        },
        "cellView": "both",
        "outputId": "c1353531-fc2a-4701-e69f-9dcfc0c4d4f9"
      },
      "source": [
        "#@markdown *Assembly and gene retrieval and processing*\n",
        "dna_url, gtf_url = urls\n",
        "try:\n",
        "    print(\" * Retrieving files...\")\n",
        "    print(dna_url)\n",
        "    print(gtf_url)\n",
        "    dna_path = os.path.basename(dna_url)\n",
        "    gtf_path = os.path.basename(gtf_url)\n",
        "    if not os.path.isfile(dna_path):\n",
        "        !wget $dna_url || printf \"\\n * Failed to download assembly\\n\"\n",
        "        if not os.path.isfile(dna_path):\n",
        "            raise FileNotFoundError('Assembly could not be downloaded.')\n",
        "    else:\n",
        "        print(\" - Skipping genome download\")\n",
        "    if not os.path.isfile(gtf_path):\n",
        "        !wget $gtf_url || printf \"\\n * Failed to download gtf\\n\"\n",
        "        if not os.path.isfile(gtf_path):\n",
        "            raise FileNotFoundError('GTF could not be downloaded.')\n",
        "    else:\n",
        "        print(\" - Skipping gtf download\")\n",
        "except Exception as e:\n",
        "    print(\" * Failed to retrieve files\")\n",
        "    print(\"{}\".format(e))\n",
        "else:\n",
        "    print(\" * Finished download\")\n",
        "    \n",
        "    print(\" * Reading gtf\")\n",
        "    ranges = pr.read_gtf(gtf_path)\n",
        "    print(\" - Merging and expanding intervals (this may take a while)...\", end=\"\")\n",
        "    merged = ranges.merge(strand=False)\n",
        "    sloppy = merged.slack(read_length // 2).merge(strand=False)\n",
        "    print(\"done\")\n",
        "    df = pd.DataFrame({'Original':[len(ranges)], 'Merged':[len(merged)], 'Expanded':[len(sloppy)]})\n",
        "    display(df)\n",
        "    bed_path = \"{}.read_until.bed\".format(dna_path)\n",
        "    sloppy.to_bed(bed_path)\n",
        "\n",
        "    print(\" * Output files:\")\n",
        "    print(\"   - Genome: {}\".format(os.path.abspath(dna_path)))\n",
        "    print(\"   - Bed   : {}\".format(os.path.abspath(bed_path)))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            " * Retrieving files...\n",
            "ftp://ftp.ensembl.org/pub/release-100/fasta/acanthochromis_polyacanthus/dna//Acanthochromis_polyacanthus.ASM210954v1.dna.toplevel.fa.gz\n",
            "ftp://ftp.ensembl.org/pub/release-100/gtf/acanthochromis_polyacanthus//Acanthochromis_polyacanthus.ASM210954v1.100.gtf.gz\n",
            " - Skipping genome download\n",
            "--2020-07-10 10:32:03--  ftp://ftp.ensembl.org/pub/release-100/gtf/acanthochromis_polyacanthus//Acanthochromis_polyacanthus.ASM210954v1.100.gtf.gz\n",
            "           => ‘Acanthochromis_polyacanthus.ASM210954v1.100.gtf.gz’\n",
            "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.197.76\n",
            "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.197.76|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /pub/release-100/gtf/acanthochromis_polyacanthus/ ... done.\n",
            "==> SIZE Acanthochromis_polyacanthus.ASM210954v1.100.gtf.gz ... 11876406\n",
            "==> PASV ... done.    ==> RETR Acanthochromis_polyacanthus.ASM210954v1.100.gtf.gz ... done.\n",
            "Length: 11876406 (11M) (unauthoritative)\n",
            "\n",
            "Acanthochromis_poly 100%[===================>]  11.33M  4.45MB/s    in 2.5s    \n",
            "\n",
            "2020-07-10 10:32:09 (4.45 MB/s) - ‘Acanthochromis_polyacanthus.ASM210954v1.100.gtf.gz’ saved [11876406]\n",
            "\n",
            " * Finished download\n",
            " * Reading gtf\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "npL8JRnPzBz_",
        "colab_type": "text"
      },
      "source": [
        "When the above code has finished executing, pressing play on the next step will download to your computer:\n",
        "\n",
        "1.   A reference genome (to provide to MinKNOW)\n",
        "2.   The source `.gtf` file from which target regions were produced.\n",
        "3.   A `.bed` file containing target regions to provide to MinKNOW."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GMV2I7WjzB78",
        "colab_type": "code",
        "colab": {},
        "cellView": "form"
      },
      "source": [
        "#@markdown *File download*\n",
        "print(\" * Downloading files:\")\n",
        "files.download(dna_path)\n",
        "files.download(gtf_path)\n",
        "files.download(bed_path)"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}